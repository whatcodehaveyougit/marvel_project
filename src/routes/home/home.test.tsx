import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import * as reactRedux from "react-redux";
import Home from "./home";
import configureStore from "redux-mock-store";
import * as characterData from "./testData/character-object.json";
import { jest, describe, it, expect } from "@jest/globals";

jest.mock("../../components/character-card/character-card.tsx", () => {
  return {
    __esModule: true,
    default: jest.fn(() => "CharacterComponent"),
  };
});
jest.mock("../../components/spinner/spinner.component.tsx", () => {
  return {
    __esModule: true,
    default: jest.fn(() => "Spinner"),
  };
});
jest.mock("../../components/input/input.component", () => {
  return {
    __esModule: true,
    default: jest.fn(() => "CustomInput"),
  };
});

const mockStore = configureStore([]);

jest.mock("react-redux", () => ({
  // ...jest.requireActual("react-redux"), // Preserve all original exports
  useSelector: jest.fn(),
  Provider: ({ children }: any) => children,
}));

describe("Home", () => {
  it("renders correctly with character data", () => {
    const initialState = {
      characters: {
        data: [characterData],
        isLoading: true,
        error: null,
      },
      character: {
        data: null,
        isLoading: false,
        error: null,
      },
      characterComics: {
        data: [],
        isLoading: true,
        error: null,
      },
    };
    const store = mockStore(initialState);

    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) => callback(store.getState()));
    const tree = render(
      <Provider store={store}>
        <Home />
      </Provider>
    ).asFragment();
    expect(tree).toMatchSnapshot();
  });
  it("renders spinner when data is loading", () => {
    // Optimization: Why do I have to create new state object each time here?
    // Instead of just changing some of the properties..
    const initialState = {
      characters: {
        data: null,
        isLoading: true,
        error: null,
      },
      character: {
        data: null,
        isLoading: true,
        error: null,
      },
      characterComics: {
        data: [],
        isLoading: true,
        error: null,
      },
    };
    const store = mockStore(initialState);
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) => callback(store.getState()));
    const tree = render(
      <Provider store={store}>
        <Home />
      </Provider>
    ).asFragment();
    expect(tree).toMatchSnapshot();
  });
  it("renders error correctly", () => {
    const initialState = {
      characters: {
        data: null,
        isLoading: true,
        error: null,
      },
      character: {
        data: characterData,
        isLoading: true,
        error: null,
      },
      characterComics: {
        data: [],
        isLoading: true,
        error: null,
      },
    };
    const store = mockStore(initialState);
    jest
      .spyOn(reactRedux, "useSelector")
      .mockImplementation((callback) => callback(store.getState()));
    const tree = render(
      <Provider store={store}>
        <Home />
      </Provider>
    ).asFragment();
    expect(tree).toMatchSnapshot();
  });
});
