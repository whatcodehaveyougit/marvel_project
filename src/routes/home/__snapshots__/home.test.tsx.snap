// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly with character data 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      CustomInput
    </div>
    <div>
      Spinner
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home renders error correctly 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      CustomInput
    </div>
    <div>
      Spinner
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home renders spinner when data is loading 1`] = `
<DocumentFragment>
  <div
    class="home-container"
  >
    <div
      class="search-bar-component"
    >
      CustomInput
    </div>
    <div>
      Spinner
    </div>
  </div>
</DocumentFragment>
`;
